extensions/eclipse-che-theia-user-preferences/src/browser/che-frontend-application.ts

```diff
@@ -0,0 +1,70 @@
+/**********************************************************************
+ * Copyright (c) 2021 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+import {
+  ApplicationShell,
+  FrontendApplication,
+  FrontendApplicationContribution,
+  KeybindingRegistry,
+  ShellLayoutRestorer,
+} from '@theia/core/lib/browser';
+import { CommandRegistry, ContributionProvider, MenuModelRegistry } from '@theia/core';
+import { inject, injectable, named } from 'inversify';
+
+import { CheShellLayoutRestorer } from './che-shell-layout-restorer';
+import { FrontendApplicationStateService } from '@theia/core/lib/browser/frontend-application-state';
+import { StorageServicePreferences } from './che-storage-preferences';
+
+@injectable()
+export class CheFrontendApplication extends FrontendApplication {
+  protected layoutPersistenceTimer: number | undefined;
+
+  constructor(
+    @inject(CommandRegistry) protected readonly commands: CommandRegistry,
+    @inject(MenuModelRegistry) protected readonly menus: MenuModelRegistry,
+    @inject(KeybindingRegistry) protected readonly keybindings: KeybindingRegistry,
+    @inject(CheShellLayoutRestorer) protected readonly layoutRestorer: ShellLayoutRestorer,
+    @inject(ContributionProvider)
+    @named(FrontendApplicationContribution)
+    protected readonly contributions: ContributionProvider<FrontendApplicationContribution>,
+    @inject(ApplicationShell) protected readonly _shell: ApplicationShell,
+    @inject(FrontendApplicationStateService) protected readonly stateService: FrontendApplicationStateService,
+    @inject(StorageServicePreferences) protected readonly storagePreferences: StorageServicePreferences
+  ) {
+    super(commands, menus, keybindings, layoutRestorer, contributions, _shell, stateService);
+  }
+
+  protected registerEventListeners() {
+    super.registerEventListeners();
+
+    window.addEventListener('resize', () => this.scheduleLayoutPersistence());
+    document.addEventListener('keydown', () => this.scheduleLayoutPersistence());
+    document.addEventListener('mousedown', () => this.scheduleLayoutPersistence());
+    document.addEventListener('mousemove', () => this.scheduleLayoutPersistence());
```

Comment:
Do we really need to track `mousemove` event?
I mean - what info we are going to persist at such event and if it's possible to replace the event by more target one.

Reply:
Possible drag and drop operations, such as moving the editor tabs.

Reply:
I asked because the event will request required operation too often as for me.
For example, I did a few clicks during couple of seconds and got hundreds such events

![onMouseEvent](https://user-images.githubusercontent.com/5676062/105743074-f4898480-5f44-11eb-97e1-54cba7e93066.gif)

Reply:
It is expected. Trigger for layout persistence will be activated after 2 seconds, when mouse stops.
We have similar implementation for activity tracker extension: [che-theia-activity-tracker-contribution.ts#L38](https://github.com/eclipse/che-theia/blob/master/extensions/eclipse-che-theia-activity-tracker/src/browser/che-theia-activity-tracker-contribution.ts#L38)

Moreover, this is common practice in js environment: 
https://stackoverflow.com/a/610000
https://stackoverflow.com/a/19262511
https://stackoverflow.com/a/15067924

Reply:
AFAIK the activity tracker sends request once per 1 minute, but it can be 30 times per 1 minute.
But OK if we don't have another way to detect drag and drop operations.

Reply:
We can track drag event with implementation that listens to drag enter/leave/drop events, but this mechanism will be redundant for this tiny functionality.